### DISCUSSION TITLE
Cachy Kernels on Gentoo

### DISCUSSION THREAD

**Discussion by sandikata** [CachyOS/linux-cachyos#23] [Ideas]
Hello.

I am writing here as i started development of cachyos patchset for gentoo.
You can see it on my page.

https://github.com/sandikata/ROKO__/tree/master/sys-kernel/cachyos-sources

Currently only most recent version 5.19.

But the only question is.
What is the reason default cachyos generic config not to boot on gentoo?

Regards,
Rosen Aleksandrov

**Comment by ptr1337** [comment]
Hey @sandikata 

I think the reason is that the config's are based on the arch config's.
Also with the PKGBUILD the default config gets a lot of changed, specially for the patches which are applied.
As you see the PKGBUILD's offer a really dynamic configuration.

I see you already use the *-all.patch, which makes it quite easy to apply the complete patch for gentoo.
Do you have possibilities to change config variables equal to the PKGBUILD does? as you see here:
https://github.com/CachyOS/linux-cachyos/blob/master/linux-bore/PKGBUILD#L253-L626
 
 But I really welcome a gentoo ebuild.

I will have a look into gentoo's config.

**Reply by sandikata** [reply]
Hello,

variables in the ebuild are possible. Just now is in early testing/development.
Even right now I've committed one more change by adding bore cacule and tt patches (which can be enabled by use flags)

Currently it works under hood of my gentoo with ryzenb 7 5800x with customized gentoo config.

**Reply by ptr1337** [reply]
Actually I dont really suggest to apply the scheduler each other, also some could conflict.

Yes, I think that gentoo has some different config variables to the one which arch is using.
Could you maybe upload me the config?

Regards

**Reply by sandikata** [reply]
Well it is not about to apply them at once, but just one of them, i have to check how to add conflicting flags in the ebuild and if some of schedulers it is enabled by use flag, if you try to enable another one at same time emerge just to complain about blocking use flags.

```
# Copyright 2022 Gentoo Authors
# Distributed under the terms of the GNU General Public License v2

EAPI=8
EXTRAVERSION="-cachyos"
K_SECURITY_UNSUPPORTED="1"
ETYPE="sources"
inherit kernel-2
detect_version

DESCRIPTION="CachyOS are improved kernels that improve performance and other aspects."
HOMEPAGE="https://github.com/CachyOS/linux-cachyos"
SRC_URI="${KERNEL_URI} \
		https://raw.githubusercontent.com/ptr1337/kernel-patches/master/${KV_MAJOR}.${KV_MINOR}/all/0001-cachyos-base-all.patch \
		https://raw.githubusercontent.com/ptr1337/kernel-patches/master/${KV_MAJOR}.${KV_MINOR}/sched/0001-bore.patch \
		https://raw.githubusercontent.com/ptr1337/kernel-patches/master/${KV_MAJOR}.${KV_MINOR}/sched/0001-cacULE-cachy.patch \
		https://raw.githubusercontent.com/ptr1337/kernel-patches/master/${KV_MAJOR}.${KV_MINOR}/sched/0001-tt-cachy.patch \
		https://raw.githubusercontent.com/ptr1337/kernel-patches/master/${KV_MAJOR}.${KV_MINOR}/misc/0001-high-hz.patch"

LICENSE=""
SLOT="5.19"
KEYWORDS="~amd64"
IUSE="bore cacule high-hz tt"
REQUIRED_USE="bore? ( !cacule !tt ) cacule? ( !bore !tt ) tt? ( !bore !cacule )"

DEPEND="virtual/linux-sources"
RDEPEND="${DEPEND}"
BDEPEND=""

src_prepare() {
	eapply "${DISTDIR}/0001-cachyos-base-all.patch"

	if use high-hz; then
		eapply "${DISTDIR}/0001-high-hz.patch"
	fi

	if use bore; then
		eapply "${DISTDIR}/0001-bore.patch"
	fi

	if use cacule; then
		eapply "{DISTDIR}/0001-cacULE-cachy.patch"
	fi

	if use tt; then
		eapply "{DISTDIR}/0001-tt-cachy.patch"
	fi

	eapply_user
}

```

Now improved with variables and blocking use flags.

```
#( 08/03/22@ 2:07 )( root@aleksandrov ):~
   USE="bore cacule tt" emerge -p cachyos-sources

These are the packages that would be merged, in order:

Calculating dependencies                             r

!!! Problem resolving dependencies for sys-kernel/cachyos-sources
... done!

!!! The ebuild selected to satisfy "cachyos-sources" has unmet requirements.
- sys-kernel/cachyos-sources-5.19.0::ROKO__ USE="bore cacule high-hz tt -build -symlink" ABI_X86="(64)"

  The following REQUIRED_USE flag constraints are unsatisfied:
    bore? ( !cacule !tt ) cacule? ( !bore !tt ) tt? ( !bore !cacule )
```

And it's not possible mixing of the patches.

**Comment by sandikata** [comment]
```
>>> Preparing source in /var/tmp/portage/sys-kernel/cachyos-sources-5.19.0/work/linux-5.19.0-cachyos ...
 * Applying 0001-cachyos-base-all.patch ...                                                                                            [ ok ]
 * Applying 0001-high-hz.patch ...                                                                                                     [ ok ]
 * Applying 0001-bore.patch ...                                                                                                        [ ok ]
 * BORE config applied
>>> Source prepared.
```

```
 * Default kernel config depending on selected scheduler has been applied.
 * You have to build kernel manually!
>>> sys-kernel/cachyos-sources-5.19.0 merged.
```

The result, maybe I will add and auto-build instructions soon.

**Reply by sandikata** [reply]
Also there's updated ebuild.

```
# Copyright 2022 Gentoo Authors
# Distributed under the terms of the GNU General Public License v2

EAPI=8
EXTRAVERSION="-cachyos"
K_SECURITY_UNSUPPORTED="1"
ETYPE="sources"
inherit kernel-2
detect_version

DESCRIPTION="CachyOS are improved kernels that improve performance and other aspects."
HOMEPAGE="https://github.com/CachyOS/linux-cachyos"
SRC_URI="${KERNEL_URI} \
		https://raw.githubusercontent.com/ptr1337/kernel-patches/master/${KV_MAJOR}.${KV_MINOR}/all/0001-cachyos-base-all.patch \
		https://raw.githubusercontent.com/ptr1337/kernel-patches/master/${KV_MAJOR}.${KV_MINOR}/sched/0001-bore.patch \
		https://raw.githubusercontent.com/ptr1337/kernel-patches/master/${KV_MAJOR}.${KV_MINOR}/sched/0001-tt-cachy.patch \
		https://raw.githubusercontent.com/ptr1337/kernel-patches/master/${KV_MAJOR}.${KV_MINOR}/misc/0001-high-hz.patch"

LICENSE=""
SLOT="5.19"
KEYWORDS="~amd64"
IUSE="bore high-hz tt"
REQUIRED_USE="bore? ( !tt ) tt? ( high-hz !bore )"

DEPEND="virtual/linux-sources"
RDEPEND="${DEPEND}"
BDEPEND=""

src_prepare() {
	eapply "${DISTDIR}/0001-cachyos-base-all.patch"

	if use high-hz; then
		eapply "${DISTDIR}/0001-high-hz.patch"
	fi

	if use bore; then
		eapply "${DISTDIR}/0001-bore.patch"
	fi

	if use tt; then
		eapply "${DISTDIR}/0001-tt-cachy.patch"
	fi

	eapply_user

	# prepare default config
	if use bore; then
		cp "${FILESDIR}/config-x86_64-bore" .config && elog "BORE config applied" || die
	fi

	if use tt; then
		cp "${FILESDIR}/config-x86_64-tt" .config && elog "TaskType config applied" || die
	fi
}

pkg_postinst() {
	elog "Default kernel config depending on selected scheduler has been applied."
	elog "You have to build kernel manually!"
}
```

**Comment by sandikata** [comment]
well, i can see you've fixed TT patch for 5.19

```
 * Applying 0001-cachyos-base-all-dev.patch ...                                                                                                          [ ok ]
 * Applying 0001-high-hz.patch ...                                                                                                                       [ ok ]
 * Applying 0001-tt-cachy-dev.patch ...                                                                                                                  [ ok ]
 * TaskType config applied
 ```

gentoo package is updated regarding your suggestions, not possible to mix any of schedulers at once, and config selected regarding scheduler.

**Comment by sandikata** [comment]
```
2022-08-16T08:18:12 >>> sys-kernel/cachyos-sources: 41 seconds
 * >>> Preparing to unpack ...
>>> Unpacking source...
>>> Unpacking linux-5.19.tar.xz to /var/tmp/portage/sys-kernel/cachyos-sources-5.19.1/work
 * Excluding Patch #5010_enable-additional-cpu-optimizations-for-gcc.patch ...                                                                           [ ok ]
 * Excluding Patch #5010_enable-additional-cpu-optimizations-for-gcc-4.9.patch ...                                                                       [ ok ]
 * Excluding Patch #5011_enable-cpu-optimizations-for-gcc8.patch ...                                                                                     [ ok ]
 * Excluding Patch #5012_enable-cpu-optimizations-for-gcc91.patch ...                                                                                    [ ok ]
 * Excluding Patch #5013_enable-cpu-optimizations-for-gcc10.patch ...                                                                                    [ ok ]
 * Applying patch-5.19.1.patch (-p1) ...                                                                                                                 [ ok ]
>>> Source unpacked in /var/tmp/portage/sys-kernel/cachyos-sources-5.19.1/work
>>> Preparing source in /var/tmp/portage/sys-kernel/cachyos-sources-5.19.1/work/linux-5.19.1-cachyos ...
 * Applying 0001-cachyos-base-all-dev.patch ...                                                                                                          [ ok ]
 * Applying 0001-high-hz.patch ...                                                                                                                       [ ok ]
 * Applying 0001-bore.patch ...                                                                                                                          [ ok ]
 * BORE config applied
>>> Source prepared.
```

same for BORE, in near future i will test and rest schedulers.

**Reply by ptr1337** [reply]
The *-dev patchset's are mainly thought for the latest stable rc (5,19,2 right now). Its better if you stay at stable to use the stable patchsets. The others are just for testing for the next release.

All in all it seems really good. I could add for building at gentoo the kernels to your repo ?

**Reply by sandikata** [reply]
Well, this is for 5.19.1, for 5.19.0 i stick to stable ones.
There's some work to be done but i am working on free time. :) 

> All in all it seems really good. I could add for building at gentoo the kernels to your repo ?

Yes you can.

**Reply by sandikata** [reply]
![Снимка на екрана на 2022-08-16 13-54-34](https://user-images.githubusercontent.com/884699/184862949-cacd8dce-9a26-4212-8617-b66fc972c475.png)

Here's how it looks, only 5.19.0 is stable (green) anything else is masked as unstable/testing (red)

**Reply by sandikata** [reply]
[https://github.com/sandikata/ROKO__/tree/master/sys-kernel/cachyos-sources](url)

They will be synchronized soon in there.
[https://gpo.zugaina.org/Overlays/ROKO__](url)

**Comment by sandikata** [comment]
```
#( 08/16/22@ 2:38 )( root@aleksandrov ):~
   dmesg | grep BORE
[    0.083661] BORE (Burst-Oriented Response Enhancer) CPU Scheduler modification 1.4.31.0 by Masahito Suzuki
#( 08/16/22@ 2:38 )( root@aleksandrov ):~
   zgrep CACHY /proc/config.gz
CONFIG_CACHY=y
#( 08/16/22@ 2:38 )( root@aleksandrov ):~
   zgrep WINESYNC /proc/config.gz
CONFIG_WINESYNC=m
#( 08/16/22@ 2:39 )( root@aleksandrov ):~
   uname -a
Linux aleksandrov.internal.tech 6.0.0-gentoo #1 SMP PREEMPT Tue Aug 16 13:34:33 EEST 2022 x86_64 AMD Ryzen 7 5800X 8-Core Processor AuthenticAMD GNU/Linux
#( 08/16/22@ 2:39 )( root@aleksandrov ):~
```

**Comment by ptr1337** [comment]
@sandikata 
That seems really good to me. Since gentoo users mostly compile their kernel itself, would it be possible to include the `auto-cpu-optimization.sh` which automatically set their march. I think since a ebuild is working like a PKGBUILD that would be possible.

I also dont know how many gentoo users are using zfs? If there are some may adding in the future zfs support like the current PKGBUILD has would be also a idea.

If you would maybe do a PR for the README with the links which gentoo users need would be nice! Then it is in! :)

**Comment by sandikata** [comment]
zfs it's not related just to gentoo, but would be nice to be implemented.
auto-cpu-optimization.sh i think could be added, but need more time for testing.

What you mean with PR ?

**Reply by ptr1337** [reply]
Doing a pull request at the readme to the repo's you are referring to

**Reply by ptr1337** [reply]
Yes, we have a option to enable zfs compatibility. But im not really sure if you are able to implement this into a ebuild.

**Reply by sandikata** [reply]
I do not even remember how to do a pull request :>

**Comment by sandikata** [comment]
![Снимка на екрана на 2022-08-18 10-32-47](https://user-images.githubusercontent.com/884699/185336777-e83de1aa-2aba-4341-acc0-5a5056d1ca0a.png)


5.19.2 in unstable (it works, but 5.19.0 still be stable as it uses stable patchset)

**Comment by nexuslinux-os** [comment]
This is really interesting. I'm currently working on a Linux distribution based on Arch Linux, and I've been looking into CachyOS kernels and their patches.

It's nice to see the CachyOS kernel being adapted for Gentoo as well. The ability to select schedulers and apply the appropriate kernel configuration through the ebuild is especially useful.

Is the Gentoo implementation still maintained today, or has the work moved to another repository/project?

**Comment by sandikata** [comment]
It is available for gentoo.. The origin was mine, but i gave up. Let's say lost interest. I do respect Peter, but honestly it depends on workflow and how you're set it up.
